File: 0122-best-time-to-buy-and-sell-stock-ii/0122-best-time-to-buy-and-sell-stock-ii.py
class Solution:
    def maxProfit(self, prices: List[int]) -> int:
        maxProfit = 0
        minShare = float("inf")

        for price in prices:
            if price > minShare:
                maxProfit += price - minShare
                minShare = price
            else:
                minShare = price
        
        return maxProfit
    
#Time Complexity: O(n)
#Space Complexity: O(1)

File: 0122-best-time-to-buy-and-sell-stock-ii/0122-best-time-to-buy-and-sell-stock-ii.java
class Solution {
    public int maxProfit(int[] prices) {
        int maxProfit = 0;
        int minPrice = Integer.MAX_VALUE;

        for(int price : prices){
            minPrice = Math.min(price, minPrice);

            if(price > minPrice){
                maxProfit += price - minPrice;
                minPrice = price;
            }
        }

        return maxProfit;
    }
}

//Time Complexity: O(n)
//Space Complexity: O(1)